Sources/SwiftGRPCNIO/ServerCallContexts/StreamingResponseCallContext.swift

```diff
@@ -28,15 +28,22 @@ open class StreamingResponseCallContext<ResponseMessage: Message>: ServerCallCon
 open class StreamingResponseCallContextImpl<ResponseMessage: Message>: StreamingResponseCallContext<ResponseMessage> {
   public let channel: Channel
 
-  public init(channel: Channel, request: HTTPRequestHead) {
+  /// - Parameters:
+  ///   - channel: The NIO channel the call is handled on.
+  ///   - request: The headers provided with this call.
+  ///   - errorDelegate: Provides a means for transforming status promise failures to `GRPCStatusTransformable` before
+  ///     sending them to the client.
+  ///
+  ///     Note: `errorDelegate` is not called for status promise that are `succeeded` with a non-OK status.
+  public init(channel: Channel, request: HTTPRequestHead, errorDelegate: ServerErrorDelegate?) {
     self.channel = channel
 
     super.init(eventLoop: channel.eventLoop, request: request)
 
     statusPromise.futureResult
       // Ensure that any error provided can be transformed to `GRPCStatus`, using "internal server error" as a fallback.
-      .recover { error in
-        (error as? GRPCStatusTransformable)?.asGRPCStatus() ?? .processingError
+      .recover { [weak errorDelegate] error in
+        ((errorDelegate?.transform(error) ?? error) as? GRPCStatusTransformable)?.asGRPCStatus() ?? .processingError
```

Comment:
I think we should be observing the error with the delegate here as well as transforming it.

Reply:
So far all errors passed to `ServerErrorDelegate.observe` are those that occur somewhere inside the pipeline, while the errors thrown into this (and the other) future have usually been thrown in "user-space" instead of "gRPC-library-space". I don't object to observing these errors as well (we're getting into "middleware" territory here, but that's not necessarily a bad thing), but maybe we should have a separate `observe` method to distinguish them? (That method could also take more context, given that e.g. the request's headers are known at this point.)

What do you think?

Reply:
Separate `observe` and `transform` methods sound good to me.

Not sure on naming; my current best suggestion is `{observe,transform}{Library,User}Error(_ error: Error)` (with any additional context for the user errors)

Reply:
I've tried incorporating this now; please let me know what you think.